logic/src/commonJvmAndroid/kotlin/com/wire/kalium/logic/feature/call/GlobalCallManager.kt

```diff
@@ -85,22 +85,22 @@ actual class GlobalCallManager(
         conversationClientsInCallUpdater: ConversationClientsInCallUpdater,
         kaliumConfigs: KaliumConfigs
     ): CallManager {
-        return callManagerHolder[userId] ?: CallManagerImpl(
-            calling = calling,
-            callRepository = callRepository,
-            userRepository = userRepository,
-            currentClientIdProvider = currentClientIdProvider,
-            selfConversationIdProvider = selfConversationIdProvider,
-            callMapper = callMapper,
-            messageSender = messageSender,
-            conversationRepository = conversationRepository,
-            federatedIdMapper = federatedIdMapper,
-            qualifiedIdMapper = qualifiedIdMapper,
-            videoStateChecker = videoStateChecker,
-            conversationClientsInCallUpdater = conversationClientsInCallUpdater,
-            kaliumConfigs = kaliumConfigs
-        ).also {
-            callManagerHolder[userId] = it
+        return callManagerHolder.computeIfAbsent(userId) {
+            CallManagerImpl(
+                calling = calling,
+                callRepository = callRepository,
+                userRepository = userRepository,
+                currentClientIdProvider = currentClientIdProvider,
+                selfConversationIdProvider = selfConversationIdProvider,
+                callMapper = callMapper,
+                messageSender = messageSender,
+                conversationRepository = conversationRepository,
+                federatedIdMapper = federatedIdMapper,
+                qualifiedIdMapper = qualifiedIdMapper,
+                videoStateChecker = videoStateChecker,
+                conversationClientsInCallUpdater = conversationClientsInCallUpdater,
+                kaliumConfigs = kaliumConfigs
+            )
```

Comment:
Not sure if I understand correctly the fix as `computeIfAbsent` is not thread safe AFAIK.

Also this way we will not have instance by logged in user.

Reply:
Previously, there was a process in place. It first checked if a map contained a "callManager" for a specific user ID. If not, it would create a new "callManager" and save its reference in the map. However, there was a problem. If two threads simultaneously performed this check, they could both conclude that there was no "callManager" for the user and both create a new one.

Now, we've improved this process by using a Java data structure called `ConcurrentHashMap`. With it, we use the `computeIfAbsent` method, which ensures that even if multiple threads try to create a "callManager" at the same time, it will be done in a way that is safe and does not result in duplicate "callManagers."

Reply:
you can check out what it does [here](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent-K-java.util.function.Function-)

Reply:
> as computeIfAbsent is not thread safe AFAIK.

`computeIfAbsent` in `HashMap` are not thread safe but we are using `ConcurrentHashMap` which it is

Reply:
but now with this changes, we are no longer adding instances to `callManagerHolder`

Reply:
> If the specified key is not already associated with a value, attempts to compute its value using the given mapping function and enters it into this map unless null

https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent-K-java.util.function.Function-